parser.go

```diff
@@ -7,30 +7,53 @@ package properties
 import (
 	"fmt"
 	"runtime"
+	"strings"
 )
 
 type parser struct {
 	lex *lexer
 }
 
-func parse(input string) (properties *Properties, err error) {
-	p := &parser{lex: lex(input)}
+func parse(input string, preserveFormatting bool) (properties *Properties, err error) {
+	p := &parser{lex: lex(input, preserveFormatting)}
 	defer p.recover(&err)
 
 	properties = NewProperties()
+	properties.PreserveFormatting = preserveFormatting
 	key := ""
-	comments := []string{}
+	comments := []Comment{}
 
 	for {
 		token := p.expectOneOf(itemComment, itemKey, itemEOF)
 		switch token.typ {
 		case itemEOF:
+			if preserveFormatting && (len(comments) > 0 || token.val != "") {
+				// There are comments at the end of the input that are not tied to a particular key
+				// Save these off against a special empty key when preserving formatting
+				if token.val != "" {
+					prefixIndex := 0
+					// Include leading whitespace into the prefix
+					prefixIndex = strings.Index(token.val, strings.TrimSpace(token.val))
+					prefix := token.val[0 : prefixIndex+1]
+					comment := Comment{prefix, token.val[prefixIndex+1 : len(token.val)]}
+					comments = append(comments, comment)
+				}
+				properties.c[""] = comments
+			}
 			goto done
 		case itemComment:
-			comments = append(comments, token.val)
+			prefix := "#"
+			prefixIndex := 0
+			if preserveFormatting {
+				// Include leading whitespace into the prefix
+				prefixIndex = strings.Index(token.val, strings.TrimSpace(token.val))
+				prefix = token.val[0:prefixIndex+1]
+			}
+			comment := Comment{prefix, token.val[prefixIndex+1:len(token.val)]}
+			comments = append(comments, comment)
 			continue
 		case itemKey:
-			key = token.val
+			key = strings.TrimSpace(token.val)
```

Comment:
This does not look right since whitespace parsing should only affect comments.

Reply:
This is necessary to maintain previous behavior. Previously, whitespace around the key was being ignored in lexBeforeKey. However, now the lexer doesn't know whether the whitespace in lexBeforeKey is part of a comment and needs to be preserved or whitespace before a key and needs to be ignored. So I do it here.